### PR TITLE
Add JsonSourceGenerationOptions.UseStringEnumConverter

### LINKED ISSUE #87195: JsonSourceGenerationOptionsAttribute should provide an option for enabling string enum serialization globally
> * While we can register the convert globally, there is no good way to ensure all enums are registered

_Originally posted by @terrajobst in https://github.com/dotnet/runtime/issues/79311#issuecomment-1579280774_

## API Proposal 

```C#
namespace System.Text.Json.Serialization;

public partial class JsonSourceGenerationOptionsAttribute
{
     public bool UseStringEnumConverter { get; set; } = false;
}
```

## API Usage

```C#
JsonSerializer.Serialize(new MyPoco(), MyContext.Default.MyPoco); // {"Property1":"Public","Property2":"Static","Property3":"Instance"}

public class MyPoco
{
    public BindingFlags Property1 { get; set; } = BindingFlags.Public;
    public BindingFlags Property2 { get; set; } = BindingFlags.Static;
    public BindingFlags Property3 { get; set; } = BindingFlags.Instance;
}

[JsonSourceGenerationOptions(UseStringEnumConverter = true)]
[JsonSerializable(typeof(MyPoco))]
public partial class MyContext { }
```

Related to https://github.com/dotnet/runtime/issues/81985 which tracks support for compile-time control of `JsonSerializerDefaults.Web`.

cc @halter73

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR:

* Implements the `JsonSourceGenerationOptionsAttribuet.UseStringEnumConverter` property.
* Implements the `JsonNumberEnumConverter<TEnum>` converter.

Fix #87195.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberEnumConverter.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Serialization
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberEnumConverter.cs:6]: How did this class not end up in System.Text.Json.Serialization.Converters?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberEnumConverter.cs:6]: I think that namespace is only used for internal converters, I don't think it's necessary for public stuff.

File: src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs [hunk 2]
```diff
@@ -132,5 +132,11 @@ public JsonSourceGenerationOptionsAttribute(JsonSerializerDefaults defaults)
         /// Specifies the source generation mode for types that don't explicitly set the mode with <see cref="JsonSerializableAttribute.GenerationMode"/>.
         /// </summary>
         public JsonSourceGenerationMode GenerationMode { get; set; }
+
+        /// <summary>
+        /// Instructs the source generator to default to <see cref="JsonStringEnumConverter"/>
+        /// instead of numeric serialization for all enum types encountered in its type graph.
+        /// </summary>
+        public bool UseStringEnumConverter { get; set; }
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs:140]: Does an equivalent option not make sense on JsonSerializerOptions?  My understanding is we're now trying to keep them in sync for any options that aren't specifically focused on one mode or the other.

**Review by stephentoub — APPROVED** [review]

**Comment by eiriktsarpalis** [comment]
It should be available starting with .NET 8 Preview 7

**Comment by svrooij** [comment]
@eiriktsarpalis any chance it will be backported to .NET7? Thanks for the super fast answer.

**Comment by eiriktsarpalis** [comment]
The bar for servicing .NET 7 is security issues or substantially broken .NET 7 features. You can still use the latest STJ version in .NET 7 using the NuGet package though.

**Pull request merged**